src/contracts/middleware/IndexRegistry.sol

```diff
@@ -0,0 +1,188 @@
+    bytes32[] public globalOperatorList;
+
+    // mapping of quorumNumber => list of operators registered for that quorum
+    mapping(uint8 => bytes32[]) public quorumToOperatorList;
+    // mapping of operatorId => quorumNumber => index history of that operator
+    mapping(bytes32 => mapping(uint8 => OperatorIndex[])) public operatorIdToIndexHistory;
+    // mapping of quorumNumber => history of numbers of unique registered operators
+    mapping(uint8 => OperatorIndex[]) public totalOperatorsHistory;
+
+    modifier onlyRegistryCoordinator() {
+        require(msg.sender == address(registryCoordinator), "IndexRegistry.onlyRegistryCoordinator: caller is not the registry coordinator");
+        _;
+    }
+
+    constructor(
+        IRegistryCoordinator _registryCoordinator
+    ){
+        registryCoordinator = _registryCoordinator;
+    }
+
+    /**
+     * @notice Registers the operator with the specified `operatorId` for the quorums specified by `quorumBitmap`.
+     * @param operatorId is the id of the operator that is being registered
+     * @param quorumNumbers is the quorum numbers the operator is registered for
+     * @dev access restricted to the RegistryCoordinator
+     */
+    function registerOperator(bytes32 operatorId, uint8[] memory quorumNumbers) external onlyRegistryCoordinator {
+        //add operator to operatorList
+        globalOperatorList.push(operatorId);
+
+        for (uint i = 0; i < quorumNumbers.length; i++) {
+            quorumToOperatorList[quorumNumbers[i]].push(operatorId);
+            _updateOperatorIdToIndexHistory(operatorId, quorumNumbers[i], uint32(quorumToOperatorList[quorumNumbers[i]].length - 1));
+            _updateTotalOperatorHistory(quorumNumbers[i]);
+        }
+    }
+
+    /**
+     * @notice Deregisters the operator with the specified `operatorId` for the quorums specified by `quorumBitmap`.
+     * @param operatorId is the id of the operator that is being deregistered
+     * @param quorumNumbers is the quorum numbers the operator is deregistered for
+     * @param quorumToOperatorListIndexes is an array of indexes for each quorum as witnesses for the last operators to swap for each quorum
+     * @param globalOperatorListIndex is the index of the operator that is to be removed from the list
+     * @dev access restricted to the RegistryCoordinator
+     */
+    function deregisterOperator(bytes32 operatorId, uint8[] memory quorumNumbers, uint32[] memory quorumToOperatorListIndexes, uint32 globalOperatorListIndex) external onlyRegistryCoordinator {
+        require(quorumNumbers.length == quorumToOperatorListIndexes.length, "IndexRegistry.deregisterOperator: quorumNumbers and indexes must be the same length");
+        _removeOperatorFromGlobalOperatorList(globalOperatorListIndex);  
+
+        for (uint i = 0; i < quorumNumbers.length; i++) {
+            _removeOperatorFromQuorumToOperatorList(quorumNumbers[i], quorumToOperatorListIndexes[i]);
```

Comment:
where is the validation being done on these indices?

Reply:
added